### PR TITLE
diagnostics: translation infrastructure

### PR DESCRIPTION
**Pull request by davidtwco**
An implementation of the infrastructure required to have translatable diagnostic messages. 

- Introduces a `DiagnosticMessage` type which can represent both the current non-translatable messages and identifiers for [Fluent](https://projectfluent.org/).
- Modifies current diagnostic API so that existing calls still work but `DiagnosticMessage`s can be provided too.
- Adds support for always loading a "fallback bundle" containing the English diagnostic messages, which are used when a `DiagnosticMessage::FluentIdentifier` is used in a diagnostic being emitted.
- Adds support for loading a "primary bundle" which contains the user's preferred language translation, and is used preferentially when it contains a diagnostic message being emitted. Primary bundles are loaded either from the path provided to `-Ztranslate-alternate-ftl` (for testing), or from the sysroot at `$sysroot/locale/$locale/*.ftl` given a locale with `-Ztranslate-lang` (which is parsed as a language identifier).
- Adds "diagnostic args" which enable normally-interpolated variables to be made available as variables for Fluent messages to use.
- Updates `#[derive(SessionDiagnostic)]` so that it can only be used for translatable diagnostics and update the handful of diagnostics which used the derive to be translatable.

For example, the following diagnostic...

```rust
#[derive(SessionDiagnostic)]
#[error = "E0195"]
pub struct LifetimesOrBoundsMismatchOnTrait {
    #[message = "lifetime parameters or bounds on {item_kind} `{ident}` do not match the trait declaration"]
    #[label = "lifetimes do not match {item_kind} in trait"]
    pub span: Span,
    #[label = "lifetimes in impl do not match this {item_kind} in trait"]
    pub generics_span: Option<Span>,
    pub item_kind: &'static str,
    pub ident: Ident,
}
```

...becomes...

```rust
#[derive(SessionDiagnostic)]
#[error(code = "E0195", slug = "typeck-lifetimes-or-bounds-mismatch-on-trait")]
pub struct LifetimesOrBoundsMismatchOnTrait {
    #[primary_span]
    #[label]
    pub span: Span,
    #[label = "generics-label"]
    pub generics_span: Option<Span>,
    pub item_kind: &'static str,
    pub ident: Ident,
}
```

```fluent
typeck-lifetimes-or-bounds-mismatch-on-trait =
    lifetime parameters or bounds on {$item_kind} `{$ident}` do not match the trait declaration
    .label = lifetimes do not match {$item_kind} in trait
    .generics-label = lifetimes in impl do not match this {$item_kind} in trait
```

r? @estebank 
cc @oli-obk @Manishearth

### CODE REVIEW

**Comment by davidtwco** [comment]
I'll wait for a review and feedback before fixing any CI failures, just in case any larger changes are required.

File: compiler/rustc_error_messages/src/lib.rs [hunk 1]
```diff
@@ -111,6 +111,13 @@ pub fn fluent_bundle(
     trace!(?locale);
     let mut bundle = FluentBundle::new(vec![locale]);
 
+    // Fluent diagnostics can insert directionality isolation markers around interpolated variables
+    // indicating that there may be a shift from right-to-left to left-to-right text (or
+    // vice-versa). These are disabled because they are sometimes visible in the error output, but
+    // may be worth investigating in future (for example: if type names are left-to-right and the
+    // surrounding diagnostic messages are right-to-left, then these might be helpful).
+    bundle.set_use_isolating(false);
```
Comment (Manishearth) [inline compiler/rustc_error_messages/src/lib.rs:119]: thought: yeah the issue here is that many terminals don't render this, but e.g. things like VSCode will. Might need to be a configured option.

File: compiler/rustc_macros/src/session_diagnostic.rs [hunk 2]
```diff
@@ -363,9 +355,9 @@ struct SessionDiagnosticDeriveBuilder<'a> {
     /// Slug is a mandatory part of the struct attribute as corresponds to the Fluent message that
     /// has the actual diagnostic message.
     slug: Option<(String, proc_macro::Span)>,
-    /// Error codes are a mandatory part of the struct attribute. Slugs may replace error codes
-    /// in future but it is desirable to mandate error codes until such a time.
-    code: Option<(String, proc_macro::Span)>,
+    /// Error codes are a optional part of the struct attribute - this is only set to detect
```
Comment (Manishearth) [inline compiler/rustc_macros/src/session_diagnostic.rs:358]: thought: feels like codes should be mandatory for now until we switch over wholesale to slugs at some point in the future, since we still want to generate the error-index

**Review by Manishearth — APPROVED** [review]
Overall looks quite good from an i18n standpoint! For the fluent stuff we probably want to support _directories_ of ftl files, not just single large ftl files.

You can see what the website does [here](https://github.com/rust-lang/www.rust-lang.org/tree/master/locales)

File: compiler/rustc_macros/src/session_diagnostic.rs (hunk 2 above)
Comment (davidtwco) [reply compiler/rustc_macros/src/session_diagnostic.rs:358]: I didn't do it that way because so many of our current diagnostics don't have error codes, and adding them increases the friction for converting to the derive (having to reserve an error code number makes the a pull request more likely to have conflict and then you need to write the description for the error). 

Any error that currently has a code should be ported to a derive that has that code, but I don't personally mind if the derive doesn't force codes to be added where they didn't exist before, we can still make a mapping from error codes to slugs even if not all slugs had an error code.

Longer-term, it would be nice if we could have the error code index be generated from mandatory documentation comments on derive structs.

I don't feel too strongly here though, I can drop that commit (making error codes optional) if others from @rust-lang/wg-diagnostics prefer.

**Comment by davidtwco** [comment]
> For the fluent stuff we probably want to support directories of ftl files, not just single large ftl files.

This is how I've implemented the user-requested language support, it loads every resource (`.ftl` file) in the `$sysroot/share/locale/$locale/` directory. I haven't done this with the fallback bundle just because I was using `include_str!`, but there's no reason we couldn't just add more `include_str!`s for each new `.ftl` file we want in the default bundle.

File: compiler/rustc_error_messages/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    let mut bundle = FluentBundle::new(vec![locale]);
+
+    // Fluent diagnostics can insert directionality isolation markers around interpolated variables
+    // indicating that there may be a shift from right-to-left to left-to-right text (or
+    // vice-versa). These are disabled because they are sometimes visible in the error output, but
+    // may be worth investigating in future (for example: if type names are left-to-right and the
+    // surrounding diagnostic messages are right-to-left, then these might be helpful).
+    bundle.set_use_isolating(false);
+
+    if let Some(requested_locale) = requested_locale {
```
Comment (davidtwco) [inline compiler/rustc_error_messages/src/lib.rs:121]: It occurs to me that if the user requests `en-US` we'll fail to find it and error even though that's the fallback bundle and we definitely have it, so I should special-case that here.

File: compiler/rustc_macros/src/session_diagnostic.rs (hunk 2 above)
Comment (Manishearth) [reply compiler/rustc_macros/src/session_diagnostic.rs:358]: Oh, I was under the impression codes were already mandatory. Never mind!!

Comment (davidtwco) [reply compiler/rustc_macros/src/session_diagnostic.rs:358]: Codes were only mandatory when using the derive, but it was only being used for something like a dozen diagnostics.

**Comment by oli-obk** [comment]
r? @oli-obk 

Taking over from Esteban to keep his queue from growing. Will sync separately with him on it

File: compiler/rustc_macros/src/session_diagnostic.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_macros/src/session_diagnostic.rs:358]: Yea, there is a fairly high number of calls to `span_err` without a code and we've not had a strong policy around them. This seems good to me and we can address it separately

File: compiler/rustc_errors/src/diagnostic.rs [hunk 4]
```diff
@@ -16,6 +18,28 @@ use std::hash::{Hash, Hasher};
 #[derive(Clone, Debug, PartialEq, Eq, Hash, Encodable, Decodable)]
 pub struct SuggestionsDisabled;
 
+/// Simplified version of `FluentArg` that can implement `Encodable` and `Decodable`. Collection of
+/// `DiagnosticArg` are converted to `FluentArgs` (consuming the collection) at the start of
+/// diagnostic emission.
+pub type DiagnosticArg<'source> = (Cow<'source, str>, DiagnosticArgValue<'source>);
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply compiler/rustc_errors/src/diagnostic.rs:25]: I'm deliberately using a tuple here because the `FromIterator` impl on `FluentArgs` works with this representation.

File: compiler/rustc_error_messages/src/lib.rs (hunk 1 above)
Comment (davidtwco) [reply compiler/rustc_error_messages/src/lib.rs]: I've added a `-Z` flag to control this for now.

File: src/tools/tidy/src/deps.rs [hunk 5]
```diff
@@ -228,8 +245,8 @@ const PERMITTED_DEPENDENCIES: &[&str] = &[
     "winapi-i686-pc-windows-gnu",
     "winapi-util",
     "winapi-x86_64-pc-windows-gnu",
-    // this is a false-positive: it's only used by rustfmt, but because it's enabled through a feature, tidy thinks it's used by rustc as well.
     "yansi-term",
+    // this is a false-positive: it's only used by rustfmt, but because it's enabled through a feature, tidy thinks it's used by rustc as well.
```
Comment (klensy) [inline src/tools/tidy/src/deps.rs:249]: Why this change?

Comment (davidtwco) [reply src/tools/tidy/src/deps.rs:249]: Just an oversight, I sorted the list automatically and that moved the comment and I didn't notice.

Comment (davidtwco) [reply src/tools/tidy/src/deps.rs]: Fixed now.

**Comment by davidtwco** [comment]
@bors r=oli-obk p=1 

Previous failure has been addressed.

**Comment by davidtwco** [comment]
@bors r=oli-obk p=1

Previous failure has been addressed.

**Comment by davidtwco** [comment]
Seems spurious

@bors retry

**Comment by davidtwco** [comment]
@bors r=oli-obk p=1

Rebased.

**Comment by bjorn3** [comment]
> This is how I've implemented the user-requested language support, it loads every resource (.ftl file) in the $sysroot/share/locale/$locale/ directory.

Does it fallback to the default sysroot if `--sysroot` is passed and this sysroot doesn't have translations?

**Comment by davidtwco** [comment]
> > This is how I've implemented the user-requested language support, it loads every resource (.ftl file) in the $sysroot/share/locale/$locale/ directory.
> 
> Does it fallback to the default sysroot if `--sysroot` is passed and this sysroot doesn't have translations?

It will only fail when it looks for a locale-specific directory in the sysroot and it isn't there, i.e. when the `$user_sysroot/share/locale/$locale` directory doesn't exist. 

If that directory exists but doesn't contain any Fluent resources (or even anything that all), then it won't fail, it'll just always fallback to the default locale. If that directory contains Fluent resources which are empty, then it also won't fail, it'll just fallback to the default locale.

I don't do anything different if it's a `--sysroot` from the user or the default.

**Comment by bjorn3** [comment]
> I don't do anything different if it's a --sysroot from the user or the default.

You might want to. Otherwise using xargo and other similar programs would cause translations to break as they don't add translations from the default sysroot to the created sysroot.

**Comment by est31** [comment]
Can the approach be changed to maybe auto-generate the ftl file or something? Because I really like the experience of grepping for error messages and then directly getting to the code that creates the error. Now I'd have to figure out a fluent identifier, then I'd have to replace the `-`s with `_`...

**Comment by est31** [comment]
Furthermore, it also makes contributor experience worse because you now have to edit one more file to add a new error/lint/etc.

Sorry for posting to this pull request but I couldn't find any issue/rfc/etc linked, and wouldn't know how to get the attention of stakeholders otherwise.

**Comment by est31** [comment]
I've found something that looks like it could be the "canonical" issue for this: https://github.com/rust-lang/community-localization/issues/4

It also looks like that initial plan included keeping the english error strings inside the rust source code.

**Comment by davidtwco** [comment]
@est31 Let's chat about this [in #t-compiler/wg-diagnostics](https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics/topic/translation/near/289156596) :)
